core/stats.rs

```
@@ -0,0 +1,207 @@
+pub struct AnalyzeStats {
+    /// Per-table statistics keyed by normalized table name.
+    pub tables: HashMap<String, TableStat>,
+}
+
+impl AnalyzeStats {
+    pub fn needs_refresh(&self) -> bool {
+        self.tables.is_empty()
+    }
+    /// Get the statistics for a table, if present.
+    pub fn table_stats(&self, table_name: &str) -> Option<&TableStat> {
+        let table_name = normalize_ident(table_name);
+        self.tables.get(&table_name)
+    }
+
+    /// Get or create the statistics bucket for a table.
+    pub fn table_stats_mut(&mut self, table_name: &str) -> &mut TableStat {
+        let table_name = normalize_ident(table_name);
+        self.tables.entry(table_name).or_default()
+    }
+
+    /// Remove all statistics for a table.
+    pub fn remove_table(&mut self, table_name: &str) {
+        let table_name = normalize_ident(table_name);
+        self.tables.remove(&table_name);
+    }
+
+    /// Remove statistics for a specific index on a table.
+    pub fn remove_index(&mut self, table_name: &str, index_name: &str) {
+        let table_name = normalize_ident(table_name);
+        let index_name = normalize_ident(index_name);
+        if let Some(table_stats) = self.tables.get_mut(&table_name) {
+            table_stats.index_stats.remove(&index_name);
+        }
+    }
+}
+
+/// Read sqlite_stat1 contents into an AnalyzeStats map without mutating schema.
+///
+/// Only regular B-tree tables and indexes are considered. Virtual and ephemeral
+/// tables are ignored.
+pub fn gather_sqlite_stat1(
+    conn: &Arc<Connection>,
+    schema: &Schema,
+    mv_tx: Option<(u64, TransactionMode)>,
+) -> Result<AnalyzeStats> {
+    let mut stats = AnalyzeStats::default();
+    let mut stmt = conn.prepare(STATS_QUERY)?;
+    stmt.set_mv_tx(mv_tx);
+    load_sqlite_stat1_from_stmt(stmt, schema, &mut stats)?;
+    Ok(stats)
+}
+
+/// Best-effort refresh analyze_stats on the connection's schema.
+pub fn refresh_analyze_stats(conn: &Arc<Connection>) {
```

Why is this a "best-effort" function i.e. we swallow errors?